source/building-your-own-javascript-components/index.html.md.erb

```diff
@@ -0,0 +1,283 @@
+- a [`createAll` function](#using-createall-with-your-components) that can initialise both GOV.UK Frontend and your own components
+- an [`isSupported` function](#checking-for-gov-uk-frontend-support-with-issupported) that checks if GOV.UK Frontend is supported
+
+Using GOV.UK Frontend’s features means you will not have to redevelop these features yourself, and it reduces the size of the [bundled JavaScript](../import-javascript/#import-javascript-using-a-bundler).
+
+## Using the `Component` class
+
+JavaScript components in GOV.UK Frontend share a number of behaviours. They:
+
+- check that GOV.UK Frontend is supported
+- check that the component is not already initialised on its root element
+- store the root element of the component
+
+The `Component` class implements these behaviours in an extensible way so you can focus the code of your component on its specific behaviour.
+
+### Defining static properties on components
+
+Some of the `Component`’s features are supported by static properties on the component inheriting from it. However, native static properties are not a JavaScript syntax supported by all [browsers that support `<script type=”module”>`](../import-javascript/#import-javascript).
+
+If your project’s tools are able to convert the syntax to code supported by browsers that support `<script type=”module”>`, you can write these properties as native static ones in your component’s class:
+
+```js
+class MyComponent extend Component{
+  static property = “value”
+}
+```
+
+Otherwise, you can add these properties to your component’s class after its definition:
+
+```js
+class MyComponent extends Component {}
+MyComponent.property = “value”
+```
+
+### Defining the `moduleName` property
+
+Components extending the `Component` class need to define a `moduleName` [static property](#defining-static-properties-on-components):
+
+```js
+class MyComponent extend Component{
+  static moduleName = “app-my-component”
+}
+```
+
+Doing this helps:
+- [`createAll` initialise the component](#using-createall-with-your-components)
+- support consistent error messaging
+
+### Customising initialisation of components
+
+When your component is initialised, it'll need to perform some tasks, such as adding event listeners. You can add these tasks to your component’s `constructor` after calling `super` to set up the base `Component`’s default behaviour.
+
+```js
+class MyComponent extends Component {
+  static moduleName = ‘app-my-component’
+
+  constructor($root) {
+    super($root)
+
+    // Run component specific initialisation here for example
+    this.$root.addEventListener(‘click’, this.handleClick.bind(this))
+  }
+
+  handleClick(event) {
+    // Handle click inside the component’s root
+  }
+}
+```
+
+### Accessing the root element in the component
+
+Within the component’s methods and its constructor, you can access the root element the component was initialised on using `this.$root`.
```

Comment:
Is there any way we can make this 'active voice'? Something like:

```suggestion
Within the component’s methods and its constructor, you can access the root element that initialised the component using `this.$root`.
```

Or "initialised with" or something similar.

If not, maybe we can just add a "that":

```suggestion
Within the component’s methods and its constructor, you can access the root element that the component was initialised on using `this.$root`.
```

Reply:
Yup, the relationship is that components are initialised on a root element, rather than a root element initialisng a component so I'll take the second suggestion. Makes me realise I tend to omit these 'that' for brevity and I probably shouldn't 😅